source/auth/auth.rst

```diff
@@ -1272,151 +1215,167 @@ mechanism_properties
 
-Drivers MUST NOT send a PrincipalStepRequest when performing automatic authentication
-or when there is a cached IdPServerResponse.  Drivers must instead  use ``saslStart`` with a JwtStepRequest.
-
-Speculative Authentication
-```````````````````````````````````
-Drivers MUST implement speculative authentication for MONGODB-OIDC during the ``hello`` handshake.  If there is an unexpired access token, the JwtStepRequest SASL command will be used as the speculation command.  If there is no cache value, the PrincipalStepRequest will be used as the speculation command.  The driver MUST NOT call any callbacks during speculative authentication.
+Interfaces
+``````````
 
-User Provided Callbacks
-```````````````````````
+Authenticating using the MONGODB-OIDC mechanism will require 1 or 2 round trips between the MongoDB driver and server.
+The requests from the driver and the replies from the server are described by the following interfaces which are encoded
+in the payload as octet sequences defining BSON objects:
 
-Drivers MUST allow the user to provide callbacks for token request and
-token refresh.  The driver MUST provide a way for the both callbacks to be either automatically
-canceled, or to cancel itself.  This can be as a timeout argument to the
-callback, a cancellation context passed to the callback, or some other
-language-appropriate mechanism.  The timeout duration MUST be 5 minutes,
-to account for the fact that there may be human interaction involved.
+.. code:: typescript
 
-Callbacks can be synchronous and/or asynchronous, depending on the driver
-and/or language.  Asynchronous callbacks should be preferred when other
-operations in the driver use asynchronous functions.
+  // Driver’s opening request in saslStart.
+  interface PrincipalStepRequest {
+    // Name of the OIDC user Principal.
+    n?: str;
+  }
 
-The driver MUST pass the following information to the request callback: ``IdpServerInfo``, and either a ``timeoutSeconds`` or ``timeoutContext`` object for the callback.   The signature of the callback is up to the driver's discretion, but the driver MUST ensure that, in the future, callbacks may have additional optional parameters passed to them.  An example might look like:
+Note that the principal name is optional as it may be provided by the IdP in environments where only one IdP is used.
+If given, then ``username`` provided by the user MUST be used as the Principal ``(n)``.
 
-.. code: typescript
+.. code:: typescript
 
-  function onRequest(info: IdpServerInfo, params: RequestParameters): IdpServerResponse
+  // The information used by a REQUEST_TOKEN_CALLBACK to authenticate with the Identity Provider.
+  interface IdpInfo {
+    // URL which describes the Authentication Server. This identifier should be
+    // the iss of provided access tokens, and be viable for RFC8414
+    // metadata discovery and RFC9207 identification.
+    issuer: str;
```

Comment:
I believe this should be `issuer: string;`. Applies to all uses of `str` as a type in this PR.

Reply:
Updated